Add hooks and HOCs to `lib/viewport`

sgomes opened the pull request:
#### Changes proposed in this Pull Request

* Fix a bug in listener removal in `lib/viewport`
* Add tests for `lib/viewport`
* Add helper hooks and higher-order components to `lib/viewport/react-helpers`.
* Add tests for `lib/viewport/react-helpers` hooks and HOCs.

#### Testing instructions

* Ensure that the modified components making use of the helpers still work. It may be difficult to trigger the reactive behaviour for some of them.
* Ensure the new tests pass.

CC @jsnajdr: I think I found a reasonable way of testing hook (and HOC) reactivity. What do you think?

sgomes committed:
Fix removeWithinBreakpointListener.

It wasn't removing listeners properly, causing components to be
updated after unmounting.

sgomes committed:
Add tests for lib/viewport.

sgomes committed:
Add helper hooks and higher order components to lib/viewport.

Also refactors some components to make use of the new helpers.

sgomes committed:
Add tests for lib/viewport hooks and HOCs.

blowery commented on client/lib/viewport/test/react-helpers.js:
```diff
@@ -0,0 +1,209 @@
+
+// Disable console warnings for this file.
+// eslint-disable-next-line no-console
+console.warn = jest.fn();
```
that's a global change though isn't it?

sgomes replied:
Doesn't every file run separately in Jest?

blowery replied:
just saying that this relies on jest isolating globals on test runs across each test file. is that a safe assumption?

sgomes replied:
It does seem potentially dangerous, I agree. I'll find a better way of doing this.

sgomes replied:
Done, using spies now.

blowery replied:
thanks!

jsnajdr replied:
> just saying that this relies on jest isolating globals on test runs across each test file. is that a safe assumption?

I agree with doing the mocks as if the globals were not isolated, but it's safe to assume that they are. The global is sandboxed, provided by the [`jest-environment`](https://jestjs.io/docs/en/configuration#testenvironment-string) as configured globally or per-file, and recreated for each file.

Each Jest file can specify a pragma at the beginning:
```js
/**
 * @jest-environment jsdom
 */
```
A rather unavoidable conclusion of this is that the global is different for each file, either Node-ish or browser-ish.

blowery commented on client/lib/viewport/index.js:
```diff
@@ -106,18 +108,28 @@ export function isWithinBreakpoint( breakpoint ) {
 }
 
 export function addWithinBreakpointListener( breakpoint, listener ) {
```
would you mind adding jsdoc's that explain what the type of each parameter is?

sgomes replied:
Done.

blowery commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,116 @@
+ * @returns {React.Component} The wrapped component.
+ */
+export function withBreakpoint( Wrapped, breakpoint ) {
+	return class extends React.Component {
```
This should also do the standard HoC displayName magic. Maybe you could use https://github.com/WordPress/gutenberg/tree/master/packages/compose/src/create-higher-order-component?

sgomes replied:
I couldn't use that library, since it doesn't support additional arguments passed to the mapping function, but I copied its functionality.

blowery replied:
nice, thanks!

sgomes committed:
Safely disable and restore console.warn for lib/viewport tests.

sgomes committed:
Add missing JSDoc to lib/viewport.

sgomes committed:
Improve displayNames for lib/viewport HOCs

sgomes committed:
Change lib/viewport listener API to use subscription tokens.

This allows the same listener to be used more than once.
Also reimplements `withBreakpoint` using `useBreakpoint`.

sgomes committed:
Change lib/viewport helper tests to create multiple instances.

sgomes commented:
All comments addressed, and several improvements made. Ready for another look.

sgomes committed:
Small cleanups to lib/viewport.

blowery commented on client/lib/viewport/test/react-helpers.js:
```diff
@@ -0,0 +1,218 @@
+		} );
+	} );
+
+	afterAll( () => {
```
maybe bump this up with the other lifecycle hooks at the top? I was just surprised to run into it at the end.

sgomes replied:
Well, since it's `afterAll` it made sense in my head to be at the bottom 😄But I can see how it can be unexpected. Moved!

blowery replied:
😄  Maybe it's just my old C++ tendencies showing

sgomes committed:
Move afterAll closer to other methods in lib/viewport tests.

blowery reviewed (approved):
LGTM, but I'd love @jsnajdr to look it over too.

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,120 @@
+		return function cleanup() {
+			removeWithinBreakpointListener( breakpoint, subscription );
+		};
+	}, []);
```
What if the `breakpoint` prop changes? A well-behaved hook should update the `isActive` value and also change the source of the subscription.

There is a ["subscribe to friend's online status" example](https://reactjs.org/docs/hooks-effect.html#explanation-why-effects-run-on-each-update) in React docs that implements the "resubscribe on friend prop change" behavior, but doesn't implement the `isOnline` state update when the observed friend changes.

That part is implemented in React's [`create-subscription`](https://github.com/facebook/react/blob/bb2939ccc23c9895d798f889d9c32848be43225e/packages/create-subscription/src/createSubscription.js#L70-L82) package, using `getDerivedStateFromProps`. But that's a render prop wrapper component, not a hook. I don't know what's the best way to adapt that technique to a `useSubscription( source )` hook.

sgomes replied:
Yes, I suppose I can't force `breakpoint` to be a static thing; a component might want to dynamically change what breakpoint it's listening to depending on, say, user choice.

As far as I can tell, the only change that's needed here is to make the effect hook depend on `breakpoint`, right? That'll trigger cleanup and a new subscription. Or am I missing a weird edge-case?

jsnajdr replied:
There indeed is a weird edge-case: if the `breakpoint` changes, the `isActive` value will have an old, incorrect value until the new listener actually fires for the first time.

It's not a blocker and we can fix as soon as we know how :smile: See also: https://github.com/facebook/react/issues/14988#issuecomment-468676819

sgomes replied:
In that case, adding this to the top of `useEffect` might fix it?

```js
const currentStatus = isWithinBreakpoint( breakpoint );
  if ( isActive !== currentStatus ) {
    setIsActive( currentStatus );
  }
```

sgomes replied:
Yes, that appears to be it. Added the change and a test for it!

jsnajdr replied:
> In that case, modifying useEffect like so might fix it?

It's an acceptable solution for our case, but wouldn't be good enough for a general-purpose `create-subscription-hook` 🙂  Consider the following component:
```js
function ShowBreakpoint( { breakpoint } ) {
  const isActive = useBreakpoint( breakpoint );
  return `${ breakpoint }: ${ isActive }`;
}
```
For a screen width 1000px, `<ShowBreakpoint breakpoint='<480px' />` renders
```
<480px: false
```
Updating the `breakpoint` prop to `>960px` first renders
```
>960px: false
```
and that's not true! Only an instant later, when the effect is executed, calls `setState` and the component is quickly rerendered to:
```
>960px: true
```

sgomes replied:
You're absolutely right! Thank you for pointing out yet another edge case.

I came up with a different implementation after a bunch of testing in a sandbox, and I think this one works correctly in every situation (famous last words, I know). Instead of doing the direct comparison and firing an update if it changes, it does things differently:

- The state is now a complex object containing both the breakpoint and the `isActive` status. We could keep it as two different states, but updating both would mean two renders, which I want to avoid in such a foundational hook.
- `handleBreakpointChange` sets all of the state in one go, and returns the previous state if nothing changes. Returning the previous state helps us avoid re-rendering if nothing changes, thanks to React's early bail.
- The hook now only returns the `isActive` status in state if `breakpoint` matches the one in state. If there's a mismatch, it means `breakpoint` was changed and the effect hasn't had a chance to run yet. In that case, it retrieves the current status from the viewport.

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,120 @@
+		return <Wrapped isBreakpointActive={ isActive } { ...props } />;
+	};
+
+	const { displayName = Wrapped.name || 'Component' } = Wrapped;
```
`const displayName = Wrapped.displayName || Wrapped.name || 'Component'` would be a lot more readable here.

sgomes replied:
👍

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,120 @@
+ */
+export function withBreakpoint( Wrapped, breakpoint ) {
+	return withBreakpointAux( Wrapped, breakpoint, 'WithBreakpoint' );
+}
```
What if the API was
```js
withBreakpoint( breakpoint )( Wrapped )
```
Could you use `@wordpress/compose` to create the HOC then?

It's a more common way how to create and use HOCs. See also `connect( mapStateToProps )( Wrapped )`. One of benefits is that HOC composition is easy:
```js
import { flowRight as compose } from 'lodash';

export default compose(
  withBreakpoint( breakpoint ),
  localize,
  withLocalizedMoment,
  connect( mapStateToProps )
)( Wrapped );
```

sgomes replied:
That's an excellent suggestion, thank you! 👍I hadn't considered having extra indirection there, but it totally makes sense. Done.

jsnajdr replied:
`createHigherOrderComponent` can be used now, too:
```js
const withBreakpoint = breakpoint => createHigherOrderComponent(
  WrappedComponent => props => {
    const isBreakpointActive = useBreakpoint( breakpoint );
    return <WrappedComponent { ...props } isBreakpointActive={ isBreakpointActive } />
  },
  'WithBreakpoint'
);
```

sgomes replied:
Done, thanks for the impl!

jsnajdr reviewed (commented):
I'm not enthusiastic about the new `addWithinBreakpointListener` API. There are two common patterns for adding and removing listeners:

**addListener and removeListener that relies on listener's identity**
Common in DOM APIs and in Node.js, standardized as `EventTarget`:
```js
const listener = () => { ... };
el.addEventListener( 'scroll', listener );
el.removeEventListener( 'scroll', listener );
```
To remove a listener, exactly the same listener object must be passed to the remove function. Both functions return void.

**subscribe function that returns an unsubscribe function**
Used in Redux (`store.subscribe`), in RxJS and in `React.useEffect`:
```js
const unsubscribe = subscribe( listener );
// forget the listener, remember the unsubscribe function
unsubscribe();
```

Can we choose one of the patterns and stick to it? `addWithinBreakpointListener` is now a confusing mixture of both.

sgomes committed:
Change the lib/viewport registration API to return the unsub fn.

This is much cleaner than maintaining a bunch of unsubscribe methods
and storing a pointer too.

sgomes commented on client/components/tooltip/index.jsx:
```diff
@@ -4,69 +4,66 @@
-		showDelay: PropTypes.number,
-		showOnMobile: PropTypes.bool,
-	};
+function Tooltip( props ) {
```
Huh, this isn't something I've had to do before, so I didn't know about that. Thank you!

I'll keep this as-is for now, and if we do run into a parent trying to set a `ref` on a `Tooltip`, we'll figure out what to do then.

jsnajdr replied:
I checked the `Tooltip` instances and didn't find any that would have a `ref`. It doesn't make much sense for this component anyway.

I just wanted to point out it's not such an innocent change as it seems :)

sgomes commented:
> I'm not enthusiastic about the new `addWithinBreakpointListener` API. There are two common patterns for adding and removing listeners:
> 
> **addListener and removeListener that relies on listener's identity**
> Common in DOM APIs and in Node.js, standardized as `EventTarget`:
> 
> ```js
> const listener = () => { ... };
> el.addEventListener( 'scroll', listener );
> el.removeEventListener( 'scroll', listener );
> ```
> 
> To remove a listener, exactly the same listener object must be passed to the remove function. Both functions return void.
> 
> **subscribe function that returns an unsubscribe function**
> Used in Redux (`store.subscribe`), in RxJS and in `React.useEffect`:
> 
> ```js
> const unsubscribe = subscribe( listener );
> // forget the listener, remember the unsubscribe function
> unsubscribe();
> ```
> 
> Can we choose one of the patterns and stick to it? `addWithinBreakpointListener` is now a confusing mixture of both.

You're right, it did get pretty confusing. I switched to the latter method, which neatly allows us to reduce the API surface and keep it more consistent 👍

sgomes committed:
Remove noop default from Tooltip.

sgomes commented on client/components/tooltip/index.jsx:
```diff
@@ -4,69 +4,66 @@
+			context={ props.context }
+			id={ props.id }
+			isVisible={ props.isVisible }
+			onClose={ noop }
```
Done.

sgomes commented on client/components/tooltip/index.jsx:
```diff
@@ -4,69 +4,66 @@
+	showDelay: 100,
+	position: 'top',
+	showOnMobile: false,
+};
```
Yes! I was very happy not to have to go the object destructuring route :)

sgomes commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,120 @@
+ * @returns {Boolean} The current status for the breakpoint.
+ */
+export function useBreakpoint( breakpoint ) {
+	const [ isActive, setIsActive ] = useState( isWithinBreakpoint( breakpoint ) );
```
Good point, I hadn't considered that!

sgomes committed:
Improvements to viewport hooks after review.

sgomes committed:
Change withBreakpoint API to add extra indirection.

sgomes commented:
Thank you for the excellent review as usual, @jsnajdr! 👍 

All comments addressed.

jsnajdr commented on client/lib/viewport/README.md:
```diff
@@ -42,28 +42,67 @@ class MyComponent extends React.Component {
+- `isDesktop()`: Whether the current screen size matches a desktop breakpoint (>960px).
+- `addWithinBreakpointListener( breakpoint, listener )`: Register a listener for size changes that affect the breakpoint. Returns the unsubscribe function.
+- `addIsMobileListener( listener )`: Register a listener for size changes that affect the mobile breakpoint (<480px). Returns the unsubscribe function.
+- `addIsDesktopListener( listener )`: Register a listener for size changes that affect the desktop breakpoint (>960px). Returns the unsubscribe function.
```
One more thing: `subscribeWithinBreakpoint` might be a better name than one that contains `addListener`. `addListener` suggests the other listener pattern, used commonly in DOM, the one that we don't use here :wink: What do you think?

sgomes replied:
Sounds good! I'll keep the `is` to make it clear we're talking about subscribing to a Boolean: `subscribeIsWithinBreakpoint`, `subscribeIsMobile`, `subscribeIsDesktop`.

jsnajdr commented on client/lib/viewport/index.js:
```diff
@@ -100,51 +100,84 @@ function getMediaQueryList( breakpoint ) {
-	if ( mediaQueryList && ! isServer ) {
-		mediaQueryList.addListener( evt => listener( evt.matches ) );
+	if ( ! listener ) {
+		return;
```
If we returned `noop` function here, we wouldn't need to check before calling the returned function. More pleasant API to work with.

sgomes replied:
Very good point. Done!

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,121 @@
+			if ( unsubscribe ) {
+				unsubscribe();
+			}
+		};
```
This could be much simpler if `unsubscribe` was always a callable function. `noop` in failed cases.

sgomes replied:
Yes, and we could even return `unsubscribe` itself, without needing to wrap it in an arrow function 👍

sgomes committed:
Further improvements to viewport API after review.

sgomes commented:
New comments addressed as well. Thanks again, @jsnajdr !

sgomes committed:
Solve edge case with useBreakpoint when breakpoint changes.

sgomes committed:
Fix some typos in lib/viewport README.

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,123 @@
+	subscribeIsWithinBreakpoint,
+	MOBILE_BREAKPOINT,
+	DESKTOP_BREAKPOINT,
+} from './index';
```
Such imports are usually written as `from '.'`

sgomes replied:
Done.

jsnajdr commented on client/lib/viewport/react-helpers.js:
```diff
@@ -0,0 +1,123 @@
+
+	function handleBreakpointChange( currentStatus ) {
+		setIsActive( currentStatus );
+	}
```
This function could be defined inside the `useEffect` callback. Or even omitted in favor of passing `setIsActive` directly as the listener. It's guaranteed to be identical for a given component instance.

sgomes replied:
Moved.

jsnajdr commented on client/components/tooltip/index.jsx:
```diff
@@ -4,69 +4,63 @@
  */
 import Popover from 'components/popover';
-import { isMobile } from 'lib/viewport';
+import { useMobileBreakpoint } from 'lib/viewport/react-helpers';
```
I think that `lib/viewport/react` would be a more pleasant name for the module. The `helpers` bit doesn't convey much information.

sgomes replied:
Renamed.

sgomes committed:
Tweak useBreakpoint again to ensure correctness at all times.

Changing to a different breakpoint should now produce an immediate
correct result, without needing to wait for an effect to fire.

There should be no unnecessary re-renders either using this approach.

sgomes committed:
Rename a parameter in useBreakpoint.

sgomes committed:
Use createHigherOrderComponent for lib/viewport HOCs.

jsnajdr commented on client/lib/viewport/test/react-helpers.js:
```diff
@@ -0,0 +1,268 @@
+	} );
+
+	beforeEach( () => {
+		container = document.createElement( 'div' );
```
It would be nice to unmount the mounted components and remove the container element after each test.

When writing reactive tests with `i18n-calypso` in #31043, I was getting this warning:
```
Warning: An update to Label inside a test was not wrapped in act(...).

When testing, code that causes React state updates should be wrapped into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */
      
This ensures that you're testing the behavior the user would see in the browser. Learn more at https://fb.me/react-wrap-tests-with-act
    in Label
```

It was caused by two tests interacting:
```
beforeEach( () => {
  container = document.createElement( 'div' );
  document.body.appendChild( container );
} );

afterEach( () => {
  document.body.removeChild( container );
  container = null;
} );

test( () => {
  act( () => {
    ReactDOM.render( <Label />, container );
  } );
} );

test( () => {
  // reset locale to default
  i18n.setLocale();

  act( () => {
    ReactDOM.render( <Label />, container );
  } );
}
```

It turned out that when the second test called `i18n.setLocale()`, the component from the first test was still mounted and reacted to updates!

Adding `ReactDOM.unmountComponentAtNode( container )` to `afterEach` fixed that.

sgomes replied:
Great idea! It's always better to run tests on a node that's actually attached to the DOM, but that does require the cleanup you mentioned. That's all in, now.

jsnajdr replied:
The React docs at https://reactjs.org/docs/test-utils.html#act say that the DOM node needs to attached to a document in order to receive any DOM events. We are lucky that the tests don't dispatch any.

sgomes committed:
Rename viewport/react-helpers to viewport/react.

sgomes committed:
Attach viewport react tests to the DOM.

jsnajdr reviewed (approved):
I can't find any more issues 😆

sgomes commented:
> I can't find any more issues 😆

You make that sound like a bad thing! 😃 

Thanks again for the thorough review and all the back-and-forth, @jsnajdr! I can confidently say this was much more complex than I originally thought it would be... 🐰 🕳
